### ISSUE TITLE
CacheControl privateCache should hold current state rather than replacing it

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-framework#8243]
Comment from @dhensby on https://github.com/silverstripe/silverstripe-framework/pull/8233

> I've noticed during testing that if you have a session active, the default state changes to private, must-revalidate which doesn't seem right to me when it should probably hold the current state and simply add private and must-revalidate to the current state...

> I'd also query how privateCache() works or, at least, how we make requests private when there's a session. I think we should probably only add the "private" directive instead of remove all of the developer's preferences ...

**Comment by chillu** [comment]
>  I've noticed during testing that if you have a session active, the default state changes to private, must-revalidate which doesn't seem right to me when it should probably hold the current state and simply add private and must-revalidate to the current state...

So you're referring to retaining `no-cache` and `no-store` if you called `disableCache()` before `privateCache()`? This would make states execution order sensitive, which I'm really not keen on (and I'm sure @sminnee would object to as well heh). 

Can you describe a use case in which caches behave differently with `Cache-Control: private, must-revalidate` vs. `Cache-Control: private, no-cache, no-store, must-revalidate` have relevant practical implications which developers care about? Keep in mind, this isn't the *only* way to define caching behaviour. If you want that level of control, you can opt for your own middleware.

**Comment by sminnee** [comment]
I don't think we should make any changes to this until we have a better understand of the circumstances in which people would use `$cacheControl->setMustRevalidate(false)`. The primary goal of the new API was to provide a smaller set of defaults that developers could rely on.

So what are some use-cases for `$cacheControl->setMustRevalidate(false)`?

**Comment by dhensby** [comment]
OK, let me start with a few things so we're all on the same page, here are some `Cache-Control` headers that are equivalent:

`Cache-Control: no-cache` == `Cache-Control: max-age=0 must-revalidate` == `Cache-Control: must-revalidate`; this also appears to be the same as having no cache-control header at all (the response is stored in the cache *but* it's only used if the server responds with a 304)

Also, the default cache-control header is now `Cache-Control: no-cache` (instead of the disabled state). This means every page load must hit the server but it *can* respond with a 304 if the ETag matches.

Adding the `private` directive tells proxies and CDNs they can't cache it, but the browser still can.

My expectation is this: As a developer I can set HTTP cache `max-age` directive to a value so that caches can store my responses without having to hit the server at all, even when there are sessions involved. If there is a session involved then *only* the client should be caching the responses and not a CDN or proxy.

So if I set up something like this:

```php
protected function init()
{
    parent::init();
    HTTPCacheControlMiddleware::singleton()->setMaxAge(300);
}
```

And there's a session active, I expect a cache-control header like this: `Cache-Control: private, max-age=300, must-revalidate`

However at the moment I'd get `Cache-Control: private, must-revalidate` which (when considering server load) is almost pointless because every request has to hit the server and thus does very little to keep load down.

I agree, I don't see much point in making `no-cache, no-store, must-revalidate` (the disabled state) become `private, no-cache, no-store, must-revalidate`, but `privateCache` doesn't modify the disabled state so the point is moot.

I don't think the use of `setMustRevalidate(false)` is the point at hand, that's a distraction and I agree very few would disable that. I'm not sure how many browsers would use a stale cache without revalidating anyway so it feels like a superfluous directive that equally can be there or not and I'm happy for it to be added

**Comment by dhensby** [comment]
> When switching from STATE_DISABLED to STATE_PRIVATE, no-store and no-cache are removed - I'd call that "modifying"? Or do you mean it only modifies with $force=true in terms of priority?

`LEVEL_DISABLED` > `LEVEL_PRIVATE` so it doesn't take effect during the `applyChangeLevel` logic

>I don't see max-age removed by any state setting? The only way this layer is influencing any previous setMaxAge() calls is through setNoStore(), which isn't called from anywhere in core

I'm testing these scenarios, so I know it's behaving this way. [`privateCache()`](https://github.com/silverstripe/silverstripe-framework/blob/4.2/src/Control/Middleware/HTTPCacheControlMiddleware.php#L609-L616) calls [`setState`](https://github.com/silverstripe/silverstripe-framework/blob/4.2/src/Control/Middleware/HTTPCacheControlMiddleware.php#L286-L293) which completely replaces the state rather than combining the current state with the private one.

**Comment by chillu** [comment]
I see. The code is pretty hard to follow on this, sorry. I agree with you now :)

We could create a whitelist of state directive overrides which are affected by `setState`?
`no-cache`, `no-store`, `private`, `public`, `must-revalidate`. But *not* `max-age`, `s-max-age`, etc. If you don't know too much about HTTP caching, as a developer we recommend that you stick to `disableCache()`, which also disables browser caching through `no-store`. If you use the more low-level `privateCache()`, you need to consider if you want browser caching through max-age. The problem here is that decision is made for you implicitly by previous `setMaxAge()` calls in your own logic. We could make this more obvious by documenting `setMaxAge(0)` alongside `privateCache()` use though.

**Comment by dhensby** [comment]
I think it might be more straight forward just to have `privateCache()` call `setStateDirective('private')` (and maybe `must-revalidate` though that does come in almost every preconfigured state we have)

**Comment by sminnee** [comment]
Alright `setMaxAge()` has a much clearer everyday usecase than disabling must-revalidate, so I agree that we need to fix this.

**Comment by dhensby** [comment]
Right, I've opened a PR with some fixes. It looks like the `setState(private)` *is* working as expected 🤦‍♂️  I think the bug I was having was with the default state...

Anyway, what's clear is this needs a bit more test coverage than it has at the moment

**Comment by chillu** [comment]
@dhensby This will go into 4.2.1, not 4.2.0. We have tagged `4.2.0-beta1` three weeks ago, and that will be released as the CWP-audited stable release shortly. We can only break this code freeze if it's a critical hotfix, which doesn't seem to apply here. We'll schedule a 4.2.1 release shortly.

**Comment by dhensby** [comment]
Oh dear, is it not possible? I'd say it is a hot fix as it fixes some bad behaviour and if we release it as 4.2.0 some of the APIs I removed / changed would be crystallised and so my PR would technically be "breaking" :/

**Comment by chillu** [comment]
If your assessment is that it's a critical hotfix, then pull it into a 4.2.0-rc1 release in your UK day. I didn't think too hard about the implications when writing my last comment, you're more in the depth of this.
